### PR TITLE
Encapsulate `ProjectionMapping` as a struct

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Follow on to https://github.com/apache/arrow-datafusion/pull/8006

## Rationale for this change

As I proposed in https://github.com/apache/arrow-datafusion/pull/8006#discussion_r1380454715 I think making stucts rather than typedefs leads to easier to maintain code as the functionality is easier to find and it allows / encourages more documentation and encapsulation

This also was a good excuse for me to try and work with the code some

## What changes are included in this PR?
1. Make `ProjectionMapping` a real structure rather than a typedef
2. Add docs
3. Move some tests to the module in which their code is defined

### CODE REVIEW

File: datafusion/physical-expr/src/equivalence.rs [hunk 1]
```diff
@@ -2565,4 +2625,234 @@ mod tests {
 
         Ok(())
     }
+
+    #[test]
+    fn test_get_indices_of_matching_sort_exprs_with_order_eq() -> Result<()> {
```
Comment (alamb) [inline datafusion/physical-expr/src/equivalence.rs:2630]: This was just moved to be with the `ProjectionMapping` struct

File: datafusion/physical-expr/src/equivalence.rs [hunk 2]
```diff
@@ -42,8 +42,65 @@ use indexmap::IndexMap;
...
+/// projection.
+#[derive(Debug, Clone)]
+pub struct ProjectionMapping {
+    /// `(source expression)` --> `(target expression)`
+    /// Indices in the vector corresponds to the indices after projection.
+    inner: Vec<(Arc<dyn PhysicalExpr>, Arc<dyn PhysicalExpr>)>,
+}
+
+impl ProjectionMapping {
+    /// Constructs the mapping between a projection's input and output
```
Comment (alamb) [inline datafusion/physical-expr/src/equivalence.rs:54]: I tried to add some additional details here in the comments to make it easier to understand. The logic is the same

File: datafusion/physical-plan/src/common.rs [hunk 3]
```diff
@@ -375,38 +373,6 @@ pub fn batch_byte_size(batch: &RecordBatch) -> usize {
     batch.get_array_memory_size()
 }
 
-/// Constructs the mapping between a projection's input and output
-pub fn calculate_projection_mapping(
```
Comment (alamb) [inline datafusion/physical-plan/src/common.rs:379]: This became `ProjectionMapping::try_new`

**Comment by alamb** [comment]
FWI @ozankabak  and @mustafasrepo  -- I realize this may conflict with WIP PRs you may have, but I had it ready so I figured I would put it up for consideration

**Review by ozankabak — APPROVED** [review]
Thank you @alamb - it LGTM. We will solve the conflicts on our end, let's get this in.

**Pull request merged**